src/libAtomVM/term.c

```diff
@@ -28,23 +28,110 @@
 
 #include <ctype.h>
 #include <inttypes.h>
+#include <stdarg.h>
+#include <stddef.h>
 #include <stdio.h>
 
+//TODO use macro from utils
+#ifndef CONTAINER_OF
+#define CONTAINER_OF(ptr, type, member) \
+    ((type *) (((char *) (ptr)) - offsetof(type, member)))
+#endif
+
+struct FprintfFun
+{
+    PrinterFun base;
+    FILE *stream;
+};
+
+struct SnprintfFun
+{
+    PrinterFun base;
+    int size;
+    char *buf;
+};
+
 const term empty_tuple = 0;
 
+int fprintf_printer(PrinterFun *fun, const char *fmt, ...)
+{
+    int ret;
+
+    va_list args;
+    va_start(args, fmt);
+
+    FILE *stream = CONTAINER_OF(fun, struct FprintfFun, base)->stream;
+    ret = vfprintf(stream, fmt, args);
+
+    va_end(args);
+
+    return ret;
+}
+
+int snprintf_printer(PrinterFun *fun, const char *fmt, ...)
+{
+    int ret;
+
+    va_list args;
+    va_start(args, fmt);
+
+    struct SnprintfFun *snpf = CONTAINER_OF(fun, struct SnprintfFun, base);
+    ret = vsnprintf(snpf->buf, snpf->size, fmt, args);
+    snpf->buf += ret;
+    snpf->size -= ret;
+
+    va_end(args);
+
+    return ret;
+}
+
 void term_display(FILE *fd, term t, const Context *ctx)
 {
+    term_fprint(fd, t, ctx->global);
+}
+
+int term_fprint(FILE *stream, term t, const GlobalContext *global)
```

Comment:
I am curious what the role of the return value is from these functions, and whether the return type should be unsigned, if (as it seems) it is initialized to 0 and always incremented (presumably by a positive value)

Reply:
See above + I'm doing a little mistake here, since I am summing it carelessly. Going to add some checks to avoid adding to the sum an error signaling negative value.